# Анализ базы данных для стартапа сервис для чтения книг по подписке

**Цель проекта:** анализ базы данных для дальнейшего формирования ценностного предложения для нового продукта.


In [1]:
import pandas as pd
import sqlalchemy as sa


In [2]:
db_config = {
'user': 'praktikum_student',
'pwd': 'Sdf4$2;d-d30pp',
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432,
'db': 'data-analyst-final-project-db'
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [3]:
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})


In [4]:
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
  with engine.connect() as con:
    return pd.read_sql(sql=sa.text(query), con = con)

## Исследование таблиц

In [5]:
#изучим таблицу books
query_books = '''SELECT * FROM books LIMIT 5'''

In [6]:
get_sql_data(query_books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [7]:
query_t = ''' SELECT * FROM books '''
test = get_sql_data(query_t)
print(test.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None


Таблица `books` содержит 6 колонок и 1000 строк. Пропущенных значений нет.
Содежание колонок:
* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.





In [8]:
#изучим таблицу authors

query_authors = '''SELECT * FROM authors LIMIT 5'''
get_sql_data(query_authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [9]:
query_a = '''SELECT * FROM authors'''
test_2 = get_sql_data(query_a)
print(test_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None


Таблица `authors` содержит сведения об авторах, количество столбцов 2, количество строк - 636. Пропущенных значений нет. Содержание столбцов:
* `author_id` — идентификатор автора;
* `author` — имя автора

In [10]:
#изучим таблицу publishers

query_publ = '''SELECT * FROM publishers LIMIT 5'''
get_sql_data(query_publ)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
(get_sql_data('''SELECT * FROM publishers''')).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Таблица `publishers` содержит сведения об издательствах. Количество столбцов - 2, количество строк - 340. Пропущенных значений нет. Содержание столбцов:
* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства

In [12]:
#изучим таблицу ratings

query_ratings = '''SELECT * FROM ratings LIMIT 5'''
get_sql_data(query_ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [13]:
(get_sql_data('''SELECT * FROM ratings''')).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Таблица `ratings` содержит информацию о пользовательских оценках книг. Количество столбцов - 4, количество строк - 6456. Пропусков нет. Содержание столбцов:
* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

In [14]:
#изучим таблицу reviews
get_sql_data('''SELECT * FROM reviews LIMIT 5''')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [15]:
(get_sql_data('''SELECT * FROM reviews''')).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Таблица `reviews` содержит информацию о  о пользовательских обзорах на книги. Количество столбцов - 4, количество строк - 2793, пропущенных значений нет. Содержание столбцов:
* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, написавшего обзор;
* `text` — текст обзора

## Задание 1: Посчитайте, сколько книг вышло после 1 января 2000 года

In [16]:
books_count = '''SELECT COUNT(DISTINCT book_id)
                 FROM books
                 WHERE publication_date > '2000-01-01'; '''
books_count_ans = get_sql_data(books_count)
print('Количество книг, выпущенных после 01.01.2000г., составляет {} шт.'.format(books_count_ans['count'][0]))

Количество книг, выпущенных после 01.01.2000г., составляет 819 шт.


### Вывод
После 1 января 2000 года вышло 819 книг.

## Задание 2: Для каждой книги посчитайте количество обзоров и среднюю оценку

In [17]:
reviews_rating = '''WITH brev AS (SELECT b.book_id,
                                         b.title,
                                         COUNT(rev.review_id) AS review_amount
                                FROM books AS b
                                LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
                                GROUP BY b.book_id),
                         br AS (SELECT b.book_id,
                                      b.title,
                                      AVG(r.rating) AS avg_rating
                               FROM books AS b
                               LEFT JOIN ratings AS r ON b.book_id = r.book_id
                               GROUP BY b.book_id)
                     SELECT brev.book_id,
                            brev.title,
                            brev.review_amount,
                            ROUND(br.avg_rating, 2) AS avg_rating
                     FROM brev
                     INNER JOIN br ON brev.book_id = br.book_id
                     GROUP BY brev.book_id, brev.title, brev.review_amount, br.avg_rating
                     ORDER BY brev.book_id;'''
data_reviews_ratings = get_sql_data(reviews_rating)
display(data_reviews_ratings)

book_id                                              title  \
0          1                                       'Salem's Lot   
1          2                 1 000 Places to See Before You Die   
2          3  13 Little Blue Envelopes (Little Blue Envelope...   
3          4  1491: New Revelations of the Americas Before C...   
4          5                                               1776   
..       ...                                                ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997                        Xenocide (Ender's Saga  #3)   
997      998                                    Year of Wonders   
998      999                        You Suck (A Love Story  #2)   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...   

     review_amount  avg_rating  
0                2        3.67  
1                1        2.50  
2                3        4.67  
3                2        4.50  
4                4        4.00  
..             ...         ...  
995              3        3.67  
996              3        3.40  
997              4        3.20  
998              2        4.50  
999              4        3.83  

[1000 rows x 4 columns]

### Вывод

Выше представлена таблица из всех 1000 книг, с ID, названием книги, количеством обзоров и среднюю оценку для каждой книги.

## Задание 3: Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [18]:
publisher = '''WITH pb AS (SELECT publisher_id,
                                  COUNT(book_id) AS books_amount
                           FROM books
                           WHERE num_pages > 50
                           GROUP BY publisher_id
                           ORDER BY books_amount DESC
                           LIMIT 1)
                SELECT pb.publisher_id,
                       p.publisher,
                       pb.books_amount
                FROM pb
                JOIN publishers AS p ON pb.publisher_id = p.publisher_id'''
print(get_sql_data(publisher))

   publisher_id      publisher  books_amount
0           212  Penguin Books            42


### Вывод

Наибольшее число книг, толщиной более 50 страниц, выпустило издательство Penguin Books (id 212). Количество выпущенных книг - 42.

## Задание 4: Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [19]:
author_1 = '''WITH t1 AS (SELECT book_id,
                                 COUNT(rating_id) OVER (PARTITION BY book_id) AS ratings_count,
                                 rating
                          FROM ratings),
              t2 AS (SELECT b.author_id,
                            b.book_id,
                            a.author
                     FROM books AS b
                     LEFT JOIN authors AS a ON b.author_id=a.author_id)
              SELECT t2.author,
                     ROUND(AVG(t1.rating), 3) AS avg_rating_author
              FROM t1
              LEFT JOIN t2 ON t1.book_id=t2.book_id
              WHERE t1.ratings_count >= 50
              GROUP BY t2.author
              ORDER BY avg_rating_author DESC
              Limit 1'''
print(get_sql_data(author_1))

                       author  avg_rating_author
0  J.K. Rowling/Mary GrandPré              4.287


### Вывод

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPre

## Задание 5: Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [20]:
reviews = '''WITH users AS (SELECT r.username,
                                   COUNT(rev.review_id) AS reviews_amount
                            FROM (SELECT username,
                                         COUNT(rating_id)
                                  FROM ratings
                                  GROUP BY username
                                  HAVING COUNT(rating_id) > 48) AS r
                            LEFT JOIN reviews AS rev ON r.username = rev.username
                            GROUP BY r.username)
              SELECT AVG(reviews_amount)
              FROM users;'''

print('Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, составляет {}'.format(get_sql_data(reviews)['avg'][0]))

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, составляет 24.0


### Вывод

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - **24**

## Вывод

Проведен анализ баз данных по поставленным задачам.
Основные выводы:

1. После 01.01.2000г. вышло **819 книг**;
2. Для каждой книги рассчитаны количество обзоров и средняя оценка. Результаты сохранены в датасете **data_reviews_ratings**;
3. Издательство, выпустившее наибольшее количество книг толщиной больше 50 страниц - **Penguin Books** (id 212, количество выпущенных книг 42);
4. Автором с самой высокой средней оценкой книг является J**.K. Rowling/Mary GrandPre** (учитывались только книги с количеством оценок 50 и выше);
5. Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - **24**